gripper/firmware/motor_hardware_g.c

```diff
@@ -0,0 +1,377 @@
+ * per full tick, so the velocity is ~(0.6552/count) um/us or ~(655.2/count) mm/s
+ * TIM2 is counting both rising and falling edges from channel A + B.
+ * so this timer uses a DIV4 prescaller to count only complete ticks
+ *
+ * The timer overflows 3 ms after an tick is received. This means any encoder
+ * movement below 333Hz, which is about 0.2184 mm/s
+ * is rejected and that we can safely assume the encoder has stopped moving.
+ **/
+void TIM4_EncoderGSpeed_Init(void) {
+    TIM_SlaveConfigTypeDef sSlaveConfig = {0};
+    TIM_MasterConfigTypeDef sMasterConfig = {0};
+    TIM_IC_InitTypeDef sConfigIC = {0};
+    htim4.Instance = TIM4;
+    htim4.Init.Prescaler =
+        calc_prescaler(SystemCoreClock, GRIPPER_ENCODER_SPEED_TIMER_FREQ);
+    htim4.Init.CounterMode = TIM_COUNTERMODE_UP;
+    htim4.Init.Period = 3000;  // timer overflows after 3 ms
+    htim4.Init.ClockDivision = TIM_CLOCKDIVISION_DIV1;
+    htim4.Init.AutoReloadPreload = TIM_AUTORELOAD_PRELOAD_DISABLE;
+    if (HAL_TIM_Base_Init(&htim4) != HAL_OK) {
+        Error_Handler();
+    }
+    /* Initialize Input Capture mode */
+    if (HAL_TIM_IC_Init(&htim4) != HAL_OK) {
+        Error_Handler();
+    }
+    /* Configure TIM4 in Slave mode */
+    sSlaveConfig.SlaveMode = TIM_SLAVEMODE_RESET;
+    sSlaveConfig.InputTrigger =
+        TIM_TS_ITR1;  // input trigger uses encoder clock
+    if (HAL_TIM_SlaveConfigSynchro(&htim4, &sSlaveConfig) != HAL_OK) {
+        Error_Handler();
+    }
+    sMasterConfig.MasterOutputTrigger = TIM_TRGO_RESET;
+    sMasterConfig.MasterSlaveMode = TIM_MASTERSLAVEMODE_DISABLE;
+    if (HAL_TIMEx_MasterConfigSynchronization(&htim4, &sMasterConfig) !=
+        HAL_OK) {
+        Error_Handler();
+    }
+    /* Initialize TIM4 input capture channel */
+    sConfigIC.ICPolarity = TIM_INPUTCHANNELPOLARITY_RISING;
+    sConfigIC.ICSelection = TIM_ICSELECTION_TRC;
+    sConfigIC.ICPrescaler = TIM_ICPSC_DIV4;
+    sConfigIC.ICFilter = 0;
+    if (HAL_TIM_IC_ConfigChannel(&htim4, &sConfigIC, TIM_CHANNEL_1) != HAL_OK) {
+        Error_Handler();
+    }
+    /*
+     * Only counter overflow/underflow generates an update interrupt.
+     * This makes sure the input capture (when the encoder is moving) does not
+     * trigger an update event!
+     */
+    __HAL_TIM_URS_ENABLE(&htim4);
+    /* The update event of the enable timer is interrupted */
+    __HAL_TIM_ENABLE_IT(&htim4, TIM_IT_UPDATE);
+    /* Start the input capture measurement */
+    HAL_TIM_IC_Start_IT(&htim4, TIM_CHANNEL_1);
+}
+
+void HAL_TIM_Encoder_MspInit(TIM_HandleTypeDef *htim) {
+    if (htim == &htim2) {
+        /* Peripheral clock enable */
+        __HAL_RCC_TIM2_CLK_ENABLE();
+        /* TIM2 interrupt Init */
+        HAL_NVIC_SetPriority(TIM2_IRQn, 7, 0);
+        HAL_NVIC_EnableIRQ(TIM2_IRQn);
+    }
+}
+
+/**
+ * @brief DAC1 Initialization Function
+ * @param None
+ * @retval None
+ */
+static void DAC1_Init(void) {
+    hdac1.Instance = DAC1;
```

Comment:
I know this isn't actually new code, but while you're in these files - we should really be explicitly setting the `State`variable in these HAL . The bootloader and main application use the same RAM and anything that isn't explicitly initialized may get changed by the bootloader - [this was an issue with the TC2](https://github.com/Opentrons/opentrons-modules/pull/400).